### DISCUSSION TITLE
Loading asynchronous data only once and using it globally

### DISCUSSION THREAD

**Discussion by MatijaNovosel** [rrousselGit/riverpod#1597] [Q&A] [answered]
Hi,

Is there a way to load some data from a server once the application runs and then expose it to the entire application?
I've been reading through the documentation and I get some concepts but I am ultimately confused about some things.

The problem I'm facing is: I fetch some data that changes from the server to display on a map then I need some data that doesn't change but I load it once the application runs, like a store of user data.

I defined a `provider` as follows:

```dart
final stationsProvider = FutureProvider.autoDispose<List<Station>>((_) async {
  var stations = await getStations();
  return stations;
});
```
The bus stations don't change but I need them on application load then access them afterwards on some parts of the application. The problem is that to get the data from a widget I need to repeat the process of calling `when` then chaining another `when` inside to get my map data, right? But how do I access the data without trying to load it again in a widget?

**Edit**: The only solution I've come up with for now is combining the providers into one and then using it after loading the data and it seems to be working but it feels like a waste:

```dart
FutureProvider<Tuple<List<Station>, TripDetails>>((ref) async {
  final stations = await ref.watch(stationsProvider.future);
  final tripDetails = await ref.watch(tripDetailsProvider(widget.serviceCalendarId).future);
  return new Tuple<List<Station>, TripDetails>(item1: stations, item2: tripDetails);
});
```

**Comment by scopendo** [comment, marked as answer]
If I understand your question correctly, what you really want is to be able to get the value of the provider later on without it being wrapped in `AsyncValue`.

There are a few ways that you can deal with this, assuming you are happy to ignore error handling.

1) Get the AsyncValue's data value, forcing not null if that suits

```dart
  // stations will have type: List<Station>
  final stations = ref.watch(stationsProvider).value!;
```

2) Create another provider whose value is essentially the same as with approach (1).  Note that in this example I've renamed your original future provider from `stationsProvider` to `asyncStationsProvider`.  In this example, I've made the provider value non-nullable but that's just a choice depending on your situation (and confidence).

```dart
// The FutureProvider for this example is renamed asyncStationsProvider
final stationsProvider = Provider.autoDispose<List<Station>>((ref) {
  var stations = ref.watch(asyncStationsProvider);
  if (stations is AsyncData) {
    // We've got a value and we assume the data value is not null
    return stations.value!;
  } else {
    throw StateError("Programming error: Stations have not loaded");
  }
});
```

3) The last approach is to load the data separately and the set the provider's value using an override on a `ProviderScope`, most likely that at the root of the widget tree.

```dart
final stationsProvider = Provider<List<Station>>((_) {
  throw StateError("Programming error: Value must be set via a ProviderScope override");
});

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();

  final stations = await getStations();

  runApp(
     ProviderScope(
      overrides: [
        stationsProvider.overrideWithValue(stations),
      ],
      child: const MyApp(),
    ),
  );
}
```

For your case, I'd probably go with the first approach.  The third approach can be handy for access to things like `SharedPreferences` that are async because they require platform plugins.

```dart
final sharedPreferencesProvider = Provider<SharedPreferences>((_) {
  throw StateError('Value must be overridden on provider scope');
});

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  final prefs = await SharedPreferences.getInstance();

  runApp(
     ProviderScope(
      overrides: [
        sharedPreferencesProvider.overrideWithValue(prefs),
      ],
      child: const MyApp(),
    ),
  );
}
```

**Reply by MatijaNovosel** [reply]
These are pretty good suggestions, thank you very much!
I think I'll stick with number 1 for now, it works so I won't question it further but I will do some more research on my own.
